File: gitdot-auth/src/handler/refresh_session.rs
use axum::{Json, extract::State, http::StatusCode};

use gitdot_api::{endpoint::auth::refresh as api, resource::auth::AuthTokensResource};
use gitdot_axum::{ClientIp, UserAgent};
use gitdot_core::dto::RefreshSessionRequest;

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
};

pub async fn refresh_session(
    State(state): State<AppState>,
    UserAgent(user_agent): UserAgent,
    ClientIp(ip_address): ClientIp,
    Json(body): Json<api::RefreshSessionRequest>,
) -> Result<AppResponse<AuthTokensResource>, AppError> {
    let request = RefreshSessionRequest::new(body.refresh_token, user_agent, ip_address.as_deref());
    state
        .session_service
        .refresh_session(request)
        .await
        .map_err(AppError::from)
        .map(|r| AppResponse::new(StatusCode::OK, r.into_api()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/auth/refresh-session.ts
import { z } from "zod";
import { AuthTokensResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const RefreshSessionRequest = z.object({
  refresh_token: z.string(),
});
export type RefreshSessionRequest = z.infer<typeof RefreshSessionRequest>;

export const RefreshSessionResponse = AuthTokensResource;
export type RefreshSessionResponse = z.infer<typeof RefreshSessionResponse>;

export const RefreshSession = {
  path: "/auth/refresh",
  method: "POST",
  request: RefreshSessionRequest,
  response: RefreshSessionResponse,
} as const satisfies Endpoint;

export type RefreshSession = typeof RefreshSession;
